### PR TITLE
async exception issue

### LINKED ISSUE #1199: Postgres connections are returned to Pool too quickly
When a query (especially long running) is interrupted while using a connection from a Pool, that connection is almost immediately returned to the pool even if the query is still running on that connection.

Another async consumer picks up the connection out of the pool to do it's work and you get:

```
libpq: failed (another command is already in progress
```

The gist of it is:

- Create a connection pool of 1
- In another thread acquire the connection from the pool doing some long running query like 10000 inserts
- Kill that thread before it finishes
- immediately try doing some other query using that same pool

I've created a reproducable repo here that just requires stack and docker (for the postgres server):

https://github.com/codygman/persistent-postgresql-query-in-progress-repro/blob/master/src/Main.hs

### PR DESCRIPTION
**Pull request by parsonsmatt**
Before submitting your PR, check that you've:

- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Bumped the version number if there isn't an `(unreleased)` on the Changelog
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

test initially thanks to @codygman 

investigating #1199 
maybe even fixes #1199

### CODE REVIEW

**Comment by parsonsmatt** [comment]
As of 9d07728, I've investigated and found something interesting:

```haskell
rawAcquireSqlConn
    :: forall backend m
     . (MonadReader backend m, BackendCompatible SqlBackend backend)
    => Maybe IsolationLevel -> m (Acquire backend)
rawAcquireSqlConn isolation = do
    conn <- MonadReader.ask
    let rawConn :: SqlBackend
        rawConn = projectBackend conn

        getter :: T.Text -> IO Statement
        getter = getStmtConn rawConn

        beginTransaction :: IO backend
        beginTransaction = conn <$ connBegin rawConn getter isolation

        finishTransaction :: backend -> ReleaseType -> IO ()
        finishTransaction _ relType = case relType of
            ReleaseExceptionWith e -> do
                putStrLn $ "got a release exception: " <> show e
                connRollback rawConn getter
                putStrLn "rolled back transaction"
            _ -> connCommit rawConn getter

    return $ mkAcquireType beginTransaction finishTransaction
```

Note the two `putStrLn` calls in `finishTransaction`. Now, look at this ioutput:

```
λ ~/Projects/persistent/ master* stack exec -- conn-kill
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 3] start inserting 100000000 things
[Info] [main thread] [ThreadId 3] inside of thing
[Info] [main thread] [ThreadId 3] Thing #: 1000
[Info] [main thread] [ThreadId 3] Thing #: 2000
[Info] [main thread] [ThreadId 3] Thing #: 3000
[Info] [main thread] [ThreadId 3] Thing #: 4000
[Info] [main thread] [ThreadId 3] Thing #: 5000
[Info] [main thread] [ThreadId 3] Thing #: 6000
[Info] [main thread] [ThreadId 3] Thing #: 7000
[Info] [main thread] [ThreadId 3] Thing #: 8000
[Info] [main thread] [ThreadId 3] Thing #: 9000
[Info] [main thread] [ThreadId 3] Thing #: 10000
[Info] [main thread] [ThreadId 3] Thing #: 11000
[Info] [main thread] [ThreadId 3] Thing #: 12000
[Info] [main thread] [ThreadId 3] Thing #: 13000
[Info] [main thread] [ThreadId 3] Thing #: 14000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
conn-kill: libpq: failed (another command is already in progress
)
creating conn
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
```

We get `connection closed, boss` and then the exception - `libpq: failed (another command is already in progress)`. We don't ever emit `rolled back transaction`. This means the exception is occuring when we try to roll back the transaction..

This is running with the non-threaded runtime. Compiling with `-threaded`, I get the same behavior. Consistently, we get `connection closed, boss` and then an exception. Now... Let's run with some threads. With two threads, no change.

<details>
```
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N2
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 7] start inserting 100000000 things
[Info] [main thread] [ThreadId 7] inside of thing
[Info] [main thread] [ThreadId 7] Thing #: 1000
[Info] [main thread] [ThreadId 7] Thing #: 2000
[Info] [main thread] [ThreadId 7] Thing #: 3000
[Info] [main thread] [ThreadId 7] Thing #: 4000
[Info] [main thread] [ThreadId 7] Thing #: 5000
[Info] [main thread] [ThreadId 7] Thing #: 6000
[Info] [main thread] [ThreadId 7] Thing #: 7000
[Info] [main thread] [ThreadId 7] Thing #: 8000
[Info] [main thread] [ThreadId 7] Thing #: 9000
[Info] [main thread] [ThreadId 7] Thing #: 10000
[Info] [main thread] [ThreadId 7] Thing #: 11000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
conn-kill: libpq: failed (another command is already in progress
)
creating conn
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N2
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 7] start inserting 100000000 things
[Info] [main thread] [ThreadId 7] inside of thing
[Info] [main thread] [ThreadId 7] Thing #: 1000
[Info] [main thread] [ThreadId 7] Thing #: 2000
[Info] [main thread] [ThreadId 7] Thing #: 3000
[Info] [main thread] [ThreadId 7] Thing #: 4000
[Info] [main thread] [ThreadId 7] Thing #: 5000
[Info] [main thread] [ThreadId 7] Thing #: 6000
[Info] [main thread] [ThreadId 7] Thing #: 7000
[Info] [main thread] [ThreadId 7] Thing #: 8000
[Info] [main thread] [ThreadId 7] Thing #: 9000
[Info] [main thread] [ThreadId 7] Thing #: 10000
[Info] [main thread] [ThreadId 7] Thing #: 11000
[Info] [main thread] [ThreadId 7] Thing #: 12000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
creating conn
conn-kill: libpq: failed (another command is already in progress
)
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N2
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 7] start inserting 100000000 things
[Info] [main thread] [ThreadId 7] inside of thing
[Info] [main thread] [ThreadId 7] Thing #: 1000
[Info] [main thread] [ThreadId 7] Thing #: 2000
[Info] [main thread] [ThreadId 7] Thing #: 3000
[Info] [main thread] [ThreadId 7] Thing #: 4000
[Info] [main thread] [ThreadId 7] Thing #: 5000
[Info] [main thread] [ThreadId 7] Thing #: 6000
[Info] [main thread] [ThreadId 7] Thing #: 7000
[Info] [main thread] [ThreadId 7] Thing #: 8000
[Info] [main thread] [ThreadId 7] Thing #: 9000
[Info] [main thread] [ThreadId 7] Thing #: 10000
[Info] [main thread] [ThreadId 7] Thing #: 11000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
creating conn
conn-kill: libpq: failed (another command is already in progress
)
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
```
</details>

With four threads?

<details>
```
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N4
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 9] start inserting 100000000 things
[Info] [main thread] [ThreadId 9] inside of thing
[Info] [main thread] [ThreadId 9] Thing #: 1000
[Info] [main thread] [ThreadId 9] Thing #: 2000
[Info] [main thread] [ThreadId 9] Thing #: 3000
[Info] [main thread] [ThreadId 9] Thing #: 4000
[Info] [main thread] [ThreadId 9] Thing #: 5000
[Info] [main thread] [ThreadId 9] Thing #: 6000
[Info] [main thread] [ThreadId 9] Thing #: 7000
[Info] [main thread] [ThreadId 9] Thing #: 8000
[Info] [main thread] [ThreadId 9] Thing #: 9000
[Info] [main thread] [ThreadId 9] Thing #: 10000
[Info] [main thread] [ThreadId 9] Thing #: 11000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
conn-kill: creating conn
libpq: failed (another command is already in progress
)
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N4
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 9] start inserting 100000000 things
[Info] [main thread] [ThreadId 9] inside of thing
[Info] [main thread] [ThreadId 9] Thing #: 1000
[Info] [main thread] [ThreadId 9] Thing #: 2000
[Info] [main thread] [ThreadId 9] Thing #: 3000
[Info] [main thread] [ThreadId 9] Thing #: 4000
[Info] [main thread] [ThreadId 9] Thing #: 5000
[Info] [main thread] [ThreadId 9] Thing #: 6000
[Info] [main thread] [ThreadId 9] Thing #: 7000
[Info] [main thread] [ThreadId 9] Thing #: 8000
[Info] [main thread] [ThreadId 9] Thing #: 9000
[Info] [main thread] [ThreadId 9] Thing #: 10000
[Info] [main thread] [ThreadId 9] Thing #: 11000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
creating conn
conn-kill: libpq: failed (another command is already in progress
)
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N4
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 9] start inserting 100000000 things
[Info] [main thread] [ThreadId 9] inside of thing
[Info] [main thread] [ThreadId 9] Thing #: 1000
[Info] [main thread] [ThreadId 9] Thing #: 2000
[Info] [main thread] [ThreadId 9] Thing #: 3000
[Info] [main thread] [ThreadId 9] Thing #: 4000
[Info] [main thread] [ThreadId 9] Thing #: 5000
[Info] [main thread] [ThreadId 9] Thing #: 6000
[Info] [main thread] [ThreadId 9] Thing #: 7000
[Info] [main thread] [ThreadId 9] Thing #: 8000
[Info] [main thread] [ThreadId 9] Thing #: 9000
[Info] [main thread] [ThreadId 9] Thing #: 10000
[Info] [main thread] [ThreadId 9] Thing #: 11000
got a release exception: thread killed
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
rolled back transaction
close' called
statements finalized boss
connection closed, boss
creating conn
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
λ ~/Projects/persistent/ matt/test-async-exception-issue* stack exec -- conn-kill +RTS -N4
[Info] [main thread] creating table...
creating conn
NOTICE:  relation "foo" already exists, skipping
[Info] [main thread] [ThreadId 9] start inserting 100000000 things
[Info] [main thread] [ThreadId 9] inside of thing
[Info] [main thread] [ThreadId 9] Thing #: 1000
[Info] [main thread] [ThreadId 9] Thing #: 2000
[Info] [main thread] [ThreadId 9] Thing #: 3000
[Info] [main thread] [ThreadId 9] Thing #: 4000
[Info] [main thread] [ThreadId 9] Thing #: 5000
[Info] [main thread] [ThreadId 9] Thing #: 6000
[Info] [main thread] [ThreadId 9] Thing #: 7000
[Info] [main thread] [ThreadId 9] Thing #: 8000
[Info] [main thread] [ThreadId 9] Thing #: 9000
[Info] [main thread] [ThreadId 9] Thing #: 10000
[Info] [main thread] [ThreadId 9] Thing #: 11000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
creating conn
conn-kill: libpq: failed (another command is already in progress
)
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
```
</details>

It's inconsistent -  I've seen two runs where it didn't happen, and a few more where it did.

🤔 

On a "successful" run, we have this log of events:

```
got a release exception: thread killed
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
rolled back transaction
close' called
statements finalized boss
connection closed, boss
creating conn
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
```

On a failed run, it looks like this:

```
[Info] [main thread] [ThreadId 9] Thing #: 11000
[Info] [main thread] killed thread
[Info] [main thread] destroying resources
[Info] [main thread] pg_sleep
got a release exception: thread killed
close' called
statements finalized boss
connection closed, boss
conn-kill: libpq: failed (another command is already in progress
)
creating conn
[Info] [main thread] result: Right [Single {unSingle = Nothing}]
```

So the exact sequence appears to matter. I'm noticing that we call `destroying resources` *before* the `close'` is called on the bad thread. The *good* thread has `got a release exception` running before the `destroy resources` is called. 

In `e03e81693d555f3606c0f73ed61d2f87f087ccee` I remove the call. The behavior is unchanged. So that's not it.

(Note to self: investigate [`resourcet-pool`](https://www.stackage.org/haddock/lts-17.6/resourcet-pool-0.1.0.0/src/Data.Pool.Acquire.html#poolToAcquire) as a potential problem here)

**Comment by parsonsmatt** [comment]
So here's `resourcet-pool`, which was pulled out from `persistent`. I think it's buggy.

```haskell
-- | Convert a 'Pool' into an 'Acquire'.
poolToAcquire :: Pool a -> Acquire a
poolToAcquire pool = fst <$> mkAcquireType getResource freeResource
  where
    getResource = takeResource pool
    freeResource (resource, localPool) = \case
      ReleaseException -> destroyResource pool localPool resource
      _ -> putResource localPool resource
```

The docs on [`takeResource`](https://www.stackage.org/haddock/lts-17.6/resource-pool-0.2.3.2/Data-Pool.html#v:takeResource) say:

> Note that this function should be used with caution, as improper exception handling can lead to leaked resources

Which, uh, does sound like a problem!

**Comment by parsonsmatt** [comment]
And, yeah, avoiding it solves the problem.

```haskell
runSqlPool
    :: forall backend m a. (MonadUnliftIO m, BackendCompatible SqlBackend backend)
    => ReaderT backend m a -> Pool backend -> m a
runSqlPool r pconn =
    -- with (acquireSqlConnFromPool pconn) $ runReaderT r
    withRunInIO $ \runInIO ->
        withResource pconn $ \backend ->
            runInIO $ runReaderT r backend
```

With this, I don't *ever* see the problem happening.

**Comment by parsonsmatt** [comment]
Let's compare `Data.Acquire.with` and `Data.Pool.withResource`:

```haskell
with :: MonadUnliftIO m
     => Acquire a
     -> (a -> m b)
     -> m b
with (Acquire f) g = withRunInIO $ \run -> E.mask $ \restore -> do
    Allocated x free <- f restore
    res <- restore (run (g x)) `E.onException` free ReleaseException
    free ReleaseNormal
    return res

withResource ::
    (MonadBaseControl IO m)
  => Pool a -> (a -> m b) -> m b
withResource pool act = control $ \runInIO -> mask $ \restore -> do
  (resource, local) <- takeResource pool
  ret <- restore (runInIO (act resource)) `onException`
            destroyResource pool local resource
  putResource local resource
  return ret
```

This is *damn* near identitcal. So what's wrong?

**Comment by parsonsmatt** [comment]
@codygman want to test this out on your codebase? i'm going to need to backport it probably too

File: persistent/Database/Persist/Sql/Run.hs [hunk 1]
```diff
@@ -88,29 +42,67 @@ runSqlPoolWithIsolation
...
+        let getter = getStmtConn sqlBackend
+        liftIO $ connRollback sqlBackend getter
+
+-- | This function is how 'runSqlPool' and 'runSqlPoolNoTransaction' are
+-- defined. In addition to the action to be performed and the 'Pool' of
+-- conections to use, we give you the opportunity to provide three actions
+-- - initialize, afterwards, and onException.
+--
+-- @since 2.12.0.0
+runSqlPoolWithHooks
```
Comment (parsonsmatt) [inline persistent/Database/Persist/Sql/Run.hs:82]: shiny new functionality! should be equivalent in terms of power.

@merijn I think you wrote the original `Acquire`-based stuff, do you mind trying this out? Specifically the `runSqlPoolNoTransaction` stuff should be what you need.

**Comment by parsonsmatt** [comment]
@codygman has told me on Slack that this fixed his problem! 😂 🎉 

Going to merge these PRs once CI passes here, and then release the backported fixes.

**Comment by merijn** [comment]
I know what's going on! The problem isn't acquire API I added in 2.10.5. The problem is that the code mixes `forkIO` and `ResourceT`, while relying on `MonadUnliftIO` to run code in the forked thread. Which you can't do.

I can't believe I didn't notice that before in @codygman's code. I opened an issue about this on `resourcet` ages ago, but solving the problem was deemed to costly to do. So, yes, any MonadResource usage that mixes with `forkIO` is likely broken. See: https://github.com/snoyberg/conduit/issues/425

**Comment by parsonsmatt** [comment]
I'm not sure that `ResourceT` is at play here, though. There's no `ResourceT` in the relevant problem code.

File: persistent/Database/Persist/Sql/Run.hs (hunk 1 above)
Comment (merijn) [reply persistent/Database/Persist/Sql/Run.hs:82]: I know this is an *extremely* late reaction, but I finally got around to (trying to) bump my code to GHC 9.2 and 9.4.

`runSqlPoolNoTransaction` is fundamentally incompatible in doing what I was using `unsafeAcquire` for, as it incurs a `MonadUnliftIO` constraint. The `Acquire` interface (like the surviving `acquireSqlConn`) only requires `MonadResource`. There is no `MonadUnliftIO` instance for `ConduitT`, which means that `runSqlPoolNoTransaction` is unusable in large parts of my code.

**Pull request merged**